### ISSUE TITLE
Ability to set response headers / cookies after yield

### ISSUE THREAD

**Issue by imbolc** [tiangolo/fastapi#661]
It feels inconsistent that we can change headers / cookies in dependencies before `yield`, but can't do this afterwords. E.g. in the following example only `before=yield` cookie will be set:
```python
def set_cookie(response: Response):
    response.set_cookie("before", "yield")
    yield
    response.set_cookie("after", "yield")

@app.get("/")
def index(_=Depends(set_cookie)):
    return {}
```

The possibility of setting them "after request" would eliminate a need for middlewares in cases when their order is not important.

Alternatively we can at least throw an exception to make the behavior explicit.

**Comment by sm-Fifteen** [comment]
How would that work? Context manager dependencies are only closed once the response has been fully rendered, I believe, at which point it's already far too late to add headers. Is there any reason why you wouldn't be able to set both headers before yielding?

**Comment by imbolc** [comment]
> Is there any reason why you wouldn't be able to set both headers before yielding?

I found it when I tried to automatically save cookie-based session.

**Comment by tiangolo** [comment]
As @sm-Fifteen says, the section after `yield` in dependencies is executed after the response has been sent. In fact, after the background tasks have been executed. This is intentional, that's the main point of dependencies with yield, to be able to run code after the response is sent.

If you want to set cookies no matter what else happens in the path operation function, you can do it in the normal dependency or in the part before yield.

Those headers are combined with any other headers you set in your dependencies or path operation and then are returned together.

**Comment by imbolc** [comment]
Right, and if feels solid until you have `Response` as a dependency. Anyway if this stuff about background tasks is intentional it will involve even more inconsistency to do any exception for dependencies with `Response`. Thanks, for your answers, guys :)

**Comment by michitaro** [comment]
I'm guessing @imbolc were talking codes something like this.
It looks quite useful.

```python
app = FastAPI()

def get_session_store(
    response: Response,
    FASTAPI_SESSION: Optional[str] = Cookie(None),
):
    store = {}
    if FASTAPI_SESSION is not None:
        store = json.loads(FASTAPI_SESSION)
    yield store
    response.set_cookie(key='FASTAPI_SESSION', value=json.dumps(store))


@app.get
def index(store: dict = Depends(get_session_store)):
    store['n'] = 42 # store will be automatically saved in cookie

@app.get
def clear(store: dict = Depends(get_session_store)):
    store.clear() # will work
    store = {'n': 42} # will not work
```

Is it possible to do something like this?

Anyway, Thank you very much for the nice framework!
